File: js/code_06.js
const fs = require('fs')
const input = fs.readFileSync('./data/dt_06.txt', 'utf8').split(',').map(Number)

/* console
url https://adventofcode.com/2021/day/6/input
const input = document.querySelector("body > pre").innerText.split(',').map(Number)  
*/

function fish_spawn (input, days) {
    //placeholder for fish count by days till spawn
    let placeholder = Array(9).fill(0); 

    //populating days till spawn placeholder with fish count
    input.map(x => {
        const prior = placeholder[x]
        placeholder.splice(x, 1, prior + 1)// replaces 1 element at index x with proir +1
    })
    for (let i = 1; i <= days; i++) {
        // The shift() method removes the first element from 
        // an array and returns that removed element
        let spawn = placeholder.shift() 
        placeholder.push(spawn) //adding new generation of fishes
        // adding count of fishes which have spawn to 7 days till spawn
        placeholder.splice(6, 1, placeholder[6] + spawn) 
    }
    return placeholder.reduce((i, j) => i + j)
}
  
const ans1 = fish_spawn(input, 80)
const ans2 = fish_spawn(input, 256)
console.log(ans1)
console.log(ans2)

File: python/code_06.py
import numpy as np

def fish_spawn(initial_pop, days):
    placeholder = np.zeros(9, np.int64) # placeholder of fish count by days till spawn
    placeholder[:6] = np.array([np.sum(initial_pop == x) for x in range(6)])

    for _ in range(days):
        spawn = placeholder[0]
        placeholder[:-1] = placeholder[1:] # moving by one day
        # adding fishes, which spawn on current day to fishes 
        # with 7 days till next spawn:
        placeholder[6] += spawn
        placeholder[8] = spawn # new generation/born fishes

    return np.sum(placeholder)

if __name__ == '__main__':
    dt = open('./data/dt_06.txt').read().split('\n')

    input = np.array(dt[0].split(','), dtype=int)

    print(f"Answer 1: {fish_spawn(input, 80)}")
    print(f"Answer 2: {fish_spawn(input, 256)}")